src/modules/flow/netctl/netctl.c

```diff
@@ -0,0 +1,336 @@
+    if(current_state != mdata->state) {
+        r = sol_flow_send_string_packet(node,
+                    SOL_FLOW_NODE_TYPE_NETCTL_SERVICE__OUT__OUT,
+                    sol_netctl_service_state_to_str(current_state));
+        SOL_INT_CHECK_GOTO(r, < 0, error);
+
+        if (mdata->state == SOL_NETCTL_SERVICE_STATE_ONLINE) {
+             r = sol_flow_send_bool_packet(node,
+                         SOL_FLOW_NODE_TYPE_NETCTL_SERVICE__OUT__ONLINE,
+                         false);
+             SOL_INT_CHECK_GOTO(r, < 0, error);
+        }
+
+        if (current_state == SOL_NETCTL_SERVICE_STATE_ONLINE) {
+              r = sol_flow_send_bool_packet(node,
+                         SOL_FLOW_NODE_TYPE_NETCTL_SERVICE__OUT__ONLINE,
+                         true);
+             SOL_INT_CHECK_GOTO(r, < 0, error);
+        }
+
+        mdata->state = current_state;
+
+        return;
+    }
+
+error:
+    mdata->state = current_state;
+
+    SOL_WRN("service %s failed to send online state", name);
+}
+
+void
+error_cb(void *data, const struct sol_netctl_service *service,
+    unsigned int error)
+{
+    struct sol_flow_node *node = data;
+    const char *name;
+
+    name = sol_netctl_service_get_name(service);
+
+    if (name)
+        sol_flow_send_error_packet(node, ENOENT, "Service %s error is %d", name, error);
+}
+
+static struct sol_netctl_service *find_service_by_name(char *service_name)
+{
+    const struct sol_ptr_vector *service_list;
+    struct sol_netctl_service *service;
+    const char *name;
+    uint16_t i;
+
+    service_list = sol_netctl_get_services();
+
+    SOL_PTR_VECTOR_FOREACH_IDX(service_list, service, i) {
+        name = sol_netctl_service_get_name(service); 
+        if (name && strcmp(name, service_name) == 0)
+            return service;
+    }
+
+    return NULL;
+}
+
+static int
+connection_handle(struct sol_flow_node *node, void *data, bool connect)
+{
+    struct network_service_data *mdata = data;
+    struct sol_netctl_service *service;
+    int r;
+
+    service = find_service_by_name(mdata->service_name);
+    if (!service) {
+        sol_flow_send_error_packet(node, ENOENT, "Did not found service name %s", mdata->service_name);
+        return 0;
+    }
+
+    if (connect) {
+       r = sol_netctl_service_connect(service);
+       if (r < 0)
+            sol_flow_send_error_packet(node, -r, "Could not connect to service: %s: %s",
+                     mdata->service_name, sol_util_strerrora(-r));
+    } else {
+       r = sol_netctl_service_disconnect(service);
+       if (r < 0)
+            sol_flow_send_error_packet(node, -r, "Could not disconnect to service: %s: %s",
+                     mdata->service_name, sol_util_strerrora(-r));
+    }
+
+    return 0;
+}
+
+static int
+connect_service_name(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    return connection_handle(node, data, true);
```

Comment:
instead of this function, put the code that matters for `connect` here. The only "duplication" is finding the service, which is a small part of the logic. You can keep that as its own function (ie: `find_or_emit_error()` function), or copy those few lines here.